src/Draco.Compiler.Api/Syntax/ParseTree.cs

```diff
@@ -42,10 +46,234 @@ public abstract partial class ParseTree
+    }
+
+    // NOTE: This might be a good general utility somewhere else?
+    private static Position StepPositionByText(Position start, string text)
+    {
+        var currLine = start.Line;
+        var currCol = start.Column;
+        for (var i = 0; i < text.Length; ++i)
+        {
+            var ch = text[i];
+            if (ch == '\r')
+            {
+                // Either Windows or OS-X 9 style newlines
+                if (i + 1 < text.Length && text[i + 1] == '\n')
+                {
+                    // Windows-style, eat extra char
+                    ++i;
+                }
+                // Otherwise OS-X 9 style
+                ++currLine;
+                currCol = 0;
+            }
+            else if (ch == '\n')
+            {
+                // Unix-style newline
+                ++currLine;
+                currCol = 0;
+            }
+            else
+            {
+                // NOTE: We might not want to increment in all cases
+                ++currCol;
+            }
+        }
+        return new(Line: currLine, Column: currCol);
+    }
+}
+
+public abstract partial class ParseTree
+{
+    public sealed partial class Token
+    {
+        private protected override Range TranslateRelativeRange(Internal.Diagnostics.RelativeRange relativeRange)
+        {
+            if (relativeRange.RelativeTo == Internal.Diagnostics.RelativeOffset.EndOfLastElement)
+            {
+                var prevToken = this.GetPrecedingToken(this);
+                var prevEnd = prevToken?.Range.End ?? new(0, 0);
+                return new(prevEnd, relativeRange.Width);
+            }
+            if (relativeRange.RelativeTo == Internal.Diagnostics.RelativeOffset.CurrentElement)
+            {
+                return new(this.Position, relativeRange.Width);
+            }
+            throw new NotSupportedException();
+        }
+
+        protected override string ComputeTextWithoutSurroundingTrivia() => this.Text;
+
+        protected override Position ComputePosition()
+        {
+            var position = new Position(Line: 0, Column: 0);
+            // Just get the position of the preceding token, if there is one
+            // If so, we offset from the end that range by the trailing trivia
+            var precedingToken = this.GetPrecedingToken(this);
+            if (precedingToken is not null)
+            {
+                position = precedingToken.Range.End;
+                foreach (var t in precedingToken.TrailingTrivia)
+                {
+                    position = StepPositionByText(position, t.Text);
+                }
+            }
+            // Offset by leading trivia
+            foreach (var t in this.LeadingTrivia)
+            {
+                position = StepPositionByText(position, t.Text);
+            }
+            // We are done
+            return position;
+        }
+
+        protected override IEnumerable<Token> GetTokens()
+        {
+            yield return this;
```

Comment:
You really love yield :p

Reply:
Honestly I just don't know what to do on single enumerables, is a singleton array better or a single yield.

Reply:
Singleton array is more efficient for this scenario, yield lead to creating a state machine only to store a single element.  

But then it's a microoptimisation, I'd weight personal taste as more important here.